v2/tools/generator/internal/codegen/pipeline/rename_properties.go

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package pipeline
+
+import (
+	"context"
+	"strings"
+
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	"github.com/pkg/errors"
+
+	"github.com/Azure/azure-service-operator/v2/tools/generator/internal/astmodel"
+	"github.com/Azure/azure-service-operator/v2/tools/generator/internal/config"
+)
+
+// RenamePropertiesStageID is the unique identifier for this pipeline stage
+const RenamePropertiesStageID = "renameProperties"
+
+func RenameProperties(cfg *config.ObjectModelConfiguration) *Stage {
+	stage := NewStage(
+		RenamePropertiesStageID,
+		"Rename properties",
+		func(_ context.Context, state *State) (*State, error) {
+			visitor := createPropertyRenamingTypeVisitor(cfg)
+			var errs []error
+			definitions := state.Definitions()
+			modified := make(astmodel.TypeDefinitionSet)
+			for name, def := range definitions {
+				_, ok := astmodel.AsObjectType(def.Type())
+				if !ok {
+					// Not an object type, so no properties to rename
+					continue
+				}
+
+				if !cfg.IsTypeConfigured(name) {
+					// If this definition isn't configured, don't do anything
+					continue
+				}
+
+				updated, err := visitor.VisitDefinition(def, name)
+				if err != nil {
+					errs = append(errs, err)
+				} else {
+					modified.Add(updated)
+				}
+			}
+
+			if len(errs) > 0 {
+				return nil, kerrors.NewAggregate(errs)
+			}
+
+			if err := cfg.RenamePropertyTo.VerifyConsumed(); err != nil {
+				return nil, errors.Wrap(err, "verifying property rename configuration")
+			}
+
+			return state.WithDefinitions(
+					definitions.OverlayWith(modified)),
+				nil
+		})
+
+	stage.RequiresPostrequisiteStages(
+		AddStatusConditionsStageID, // Must rename other properties before we try to introduce the `Conditions` property
+		AddOperatorSpecStageID,     // Must rename other properties before we try to introduce the `OperatorSpec` property
+	)
+
+	return stage
+}
+
+func createPropertyRenamingTypeVisitor(
+	cfg *config.ObjectModelConfiguration,
+) astmodel.TypeVisitor[astmodel.InternalTypeName] {
+	builder := astmodel.TypeVisitorBuilder[astmodel.InternalTypeName]{
+		VisitObjectType: func(
+			this *astmodel.TypeVisitor[astmodel.InternalTypeName],
+			it *astmodel.ObjectType,
+			ctx astmodel.InternalTypeName,
+		) (astmodel.Type, error) {
+			return renamePropertiesInObjectType(ctx, it, cfg)
+		},
+	}
+
+	return builder.Build()
+}
+
+func renamePropertiesInObjectType(
+	typeName astmodel.InternalTypeName,
+	ot *astmodel.ObjectType,
+	cfg *config.ObjectModelConfiguration,
+) (astmodel.Type, error) {
+
+	properties := make([]*astmodel.PropertyDefinition, 0, ot.Properties().Len())
+	err := ot.Properties().ForEachError(
+		func(prop *astmodel.PropertyDefinition) error {
+			name, err := cfg.RenamePropertyTo.Lookup(typeName, prop.PropertyName())
+			if err != nil {
```

Comment:
I was talking with Artur and Christian, and realized that `errors.New` and similar `errors` functions (such as `Wrapf`) capture stacktraces that we're not logging.

I think in this case we don't actually wrap the returned error before we check it here (it's returned directly from `Lookup` which doesn't use `errors.New` or any similar such things.

It does seem like it might be an easy mistake to make to end up wrapping an error that's being used for control-flow, triggering a ton of stacktraces to be gathered. I'm wondering if we should think of other ways to implement this pattern rather than using an error? Or maybe an error is fine and we should just be careful about wrapping it (I think we are in general).

Just something to think about.

Reply:
I wrote `Lookup()` to return `(value, error)` because returning a triple of `(value, bool, error)` seemed more than a bit nasty. We could change it to return `(LookupResult, error)` where `LookupResult` took care of returning the `NotFound` outcome.

Reply:
I've logged #3545 to capture this.

Reply:
Nice. Resolving this comment as you've definitely addressed it as much as appropriate in this PR.